dev-docs/RFCs/v7.1/keyframe-rfc.md

```diff
@@ -0,0 +1,80 @@
+completely up to the application. The key frame manager will not attempt to actually interpolate the data as that would require knowledge of the data being used by the application and add unnecessary complexity to the proposed API. Instead, the key frame manager will provide the application with all data needed to perform the correct interpolation:
+- Index and application data of the start key frame
+- Index and application data of the end key frame
+- The factor to use to interpolate between the start and end key frames
+
+
+The `Timeline` class will be minimally extended to support attaching what will be referred to as "animations", defined as any object with a `setTime` method. When a timelime's time value is updated, it will cascade that value down to any objects attached in this way.
+
+## Implementation
+
+A `KeyFrames` class that is constructed from an array of [time, data] pairs that represent the key frame times and application data associated with them. It will provide the following methods:
+- setKeyFrames(): Replaces the current set of key frames with a new one. Takes the same argument as the constructor
+- getStartIndex(): Returns the current start key frame index (i.e. the index of the key frame being interpolated from)
+- getEndIndex(): Returns the current end key frame index (i.e. the index of the key frame being interpolated to)
+- getStartData(): Returns the data at the current start key frame index (i.e. the data being interpolated from)
+- getEndData(): Returns the data at the current end key frame index (i.e. the data being interpolated to)
+- getFactor(): Returns a value between 0 and 1 representing the interpolation factor between the start and end key frames pair
+- setTime(): Set the current time of the key frames
+
+The `Timeline` class will be extended with two methods:
+- `attachAnimation`: takes an animation object (i.e. any object with a `setTime` method) and optionally a channel handle, and will update that object's time whenever `Timeline.setTime()` is called. Returns a handle to that animations attachment point.
+- `detachAnimation`: takes an animation attachment handle and detaches the associated animation from the timeline.
+
+
+## Example
+
+```js
+
+const kf = new KeyFrames([
+  [0, { val1: [1, 0, 1], val2: 0} ],
+  [500, { val1: [1, 1, 1], val2: 2} ],
+  [800, { val1: [0, 0, 1], val2: 1} ],
+  [1200, { val1: [0, 1, 0], val2: 4} ],
+  [1500, { val1: [1, 0, 1], val2: 5} ]
+]);
+
+const handle = timeline.attachAnimation(kf, channel);
+
+timeline.setTime(1000);
+
+kf.getStartIndex(); // => 2                            (i.e. key frame at time=800)
+kf.getEndIndex();   // => 3                            (i.e. key frame at time=1200)
+kf.getStartData()   // => { val1: [0, 0, 1], val2: 1}  (i.e. data at index 2)
+kf.getEndData()     // => { val1: [0, 1, 0], val2: 4}  (i.e. data at index 3)
+kf.getFactor();     // => 0.5                          (i.e. halfway between 800 and 1200)
```

Comment:
What will we return as the index (and as the data) when the current time is before the first keyframe or after the last keyframe? We won't be animating from or to any values in those cases.

Reply:
They will will be clamped to pairs (0, 1) and (numKeys - 2, numKeys - 1).